packages/itwin/tree-widget/src/components/trees/experimental/common/UseHierarchyFiltering.tsx

```diff
@@ -0,0 +1,193 @@
+      }
+
+      const descriptor = await Presentation.presentation.getContentDescriptor({
+        imodel,
+        rulesetOrId: {
+          id: `Hierarchy level descriptor ruleset`,
+          rules: [
+            {
+              ruleType: "Content",
+              specifications: [
+                {
+                  specType: "SelectedNodeInstances",
+                },
+              ],
+            },
+          ],
+        },
+        displayType: DefaultContentDisplayTypes.PropertyPane,
+        keys: new KeySet(inputKeys),
+      });
+      if (!descriptor) {
+        throw new Error("Failed to create content descriptor");
+      }
+
+      return { descriptor, inputKeys };
+    };
+  }, [filteringOptions, imodel]);
+
+  const getInitialFilter = useMemo(() => {
+    const currentFilter = filteringOptions?.hierarchyDetails.instanceFilter;
+    if (!currentFilter) {
+      return undefined;
+    }
+
+    return (descriptor: Descriptor) => fromGenericFilter(descriptor, currentFilter);
+  }, [filteringOptions]);
+
+  const filteringDialog = (
+    <PresentationInstanceFilterDialog
+      imodel={imodel}
+      isOpen={!!filteringOptions}
+      onApply={(info) => {
+        if (!filteringOptions) {
+          return;
+        }
+        filteringOptions.hierarchyDetails.setInstanceFilter(toGenericFilter(info));
+        setFilteringOptions(undefined);
+      }}
+      onClose={() => {
+        setFilteringOptions(undefined);
+      }}
+      propertiesSource={propertiesSource}
+      initialFilter={getInitialFilter}
+      filterResultsCountRenderer={(filter) => {
+        if (!filteringOptions) {
+          return null;
+        }
+
+        return (
+          <MatchingInstancesCount
+            filter={filter}
+            hierarchyLevelDetails={filteringOptions.hierarchyDetails}
+            defaultHierarchyLevelSizeLimit={defaultHierarchyLevelSizeLimit}
+          />
+        );
+      }}
+    />
+  );
+
+  return {
+    onFilterClick,
+    filteringDialog,
+  };
+}
+
+interface MatchingInstancesCountProps {
+  filter: PresentationInstanceFilterInfo;
+  hierarchyLevelDetails: HierarchyLevelDetails;
+  defaultHierarchyLevelSizeLimit: number;
+}
+
+function MatchingInstancesCount({ filter, defaultHierarchyLevelSizeLimit, hierarchyLevelDetails }: MatchingInstancesCountProps) {
+  const { value } = useDebouncedAsyncValue(
+    useCallback(async () => {
+      const instanceFilter = toGenericFilter(filter);
+      try {
+        const instanceKeys = await collectInstanceKeys(
+          hierarchyLevelDetails.getInstanceKeysIterator({
+            instanceFilter,
+            hierarchyLevelSizeLimit: hierarchyLevelDetails.sizeLimit ?? defaultHierarchyLevelSizeLimit,
+          }),
+        );
+        return `Current filter matching instances count: ${instanceKeys.length}`;
+      } catch (e) {
+        if (e instanceof RowsLimitExceededError) {
+          return `Current filter exceeds instances count of ${e.limit}`;
+        }
```

Comment:
I think here we should either rethrow the error or return a message saying there was an error getting the instances count.

Reply:
Updated it to return error message